metricbeat/module/kvm/dommemstat/dommemstat.go

```diff
@@ -0,0 +1,140 @@
+package dommemstat
+
+import (
+	"errors"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/metricbeat/mb"
+
+	"github.com/digitalocean/go-libvirt"
+	"github.com/digitalocean/go-libvirt/libvirttest"
+)
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each host defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet("kvm", "dommemstat", New)
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	mb.BaseMetricSet
+	separatedURI []string
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Experimental("The kvm dommemstat metricset is experimental.")
+
+	config := struct{}{}
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+
+	return &MetricSet{
+		BaseMetricSet: base,
+		separatedURI:  strings.Split(base.HostData().URI, "://"),
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) {
+
+	var (
+		c   net.Conn
+		err error
+	)
+
+	if m.separatedURI[0] == "test" {
+		// when running tests, a mock Libvirt server is used
+		c = libvirttest.New()
+	} else {
+		c, err = net.DialTimeout(m.separatedURI[0], m.separatedURI[1], 2*time.Second)
+		if err != nil {
+			report.Error(err)
+		}
+	}
+
+	l := libvirt.New(c)
+	if err := l.Connect(); err != nil {
+		report.Error(err)
+	}
+
+	domains, err := l.Domains()
+	if err != nil {
+		report.Error(err)
+	}
+
+	for _, d := range domains {
+		l.DomainMemoryStats(d, 8, 0)
+
+		gotDomainMemoryStats, err := l.DomainMemoryStats(d, 8, 0)
```

Comment:
What are these 8 an 0? Would it make sense to set from constants?

Reply:
8 -> Maximum stats that you want
0 -> Flags that you can pass
Actually it would be better to change to 11 since we have 11 possible stats and to set constants.